test/zdtm/static/rseq.c

```diff
@@ -0,0 +1,190 @@
+{
+	return syscall(__NR_rseq, rseq_abi, rseq_len, flags, sig);
+}
+
+static void register_thread(void)
+{
+	int rc;
+	rc = sys_rseq(&__rseq_abi, sizeof(struct rseq), 0, RSEQ_SIG);
+	if (rc) {
+		fail("Failed to register rseq");
+		exit(1);
+	}
+}
+
+static void unregister_thread(void)
+{
+	int rc;
+	rc = sys_rseq(&__rseq_abi, sizeof(struct rseq), RSEQ_FLAG_UNREGISTER, RSEQ_SIG);
+	if (rc) {
+		fail("Failed to unregister rseq");
+		exit(1);
+	}
+}
+
+static void check_thread(void)
+{
+	int rc;
+	rc = sys_rseq(&__rseq_abi, sizeof(struct rseq), 0, RSEQ_SIG);
+	if (!(rc && errno == EBUSY)) {
+		fail("Failed to check rseq %d", rc);
+		exit(1);
+	}
+}
+
+#define RSEQ_ACCESS_ONCE(x) (*(__volatile__ __typeof__(x) *)&(x))
+
+static int rseq_addv(intptr_t *v, intptr_t count, int cpu)
+{
+	/* clang-format off */
+	__asm__ __volatile__ goto(
+		".pushsection __rseq_table, \"aw\"\n\t"
+		".balign 32\n\t"
+		"cs_obj:\n\t"
+		/* version, flags */
+		".long 0, 0\n\t"
+		/* start_ip, post_commit_ip, abort_ip */
+		".quad 1f, 2f, 4f\n\t"
+		".popsection\n\t"
+		"1:\n\t"
+		"leaq cs_obj(%%rip), %%rax\n\t"
+		"movq %%rax, %[rseq_cs]\n\t"
+		"cmpl %[cpu_id], %[current_cpu_id]\n\t"
+		"jnz 4f\n\t"
+		"addq %[count], %[v]\n\t"	/* final store */
+		"2:\n\t"
+		".pushsection __rseq_failure, \"ax\"\n\t"
+		/* Disassembler-friendly signature: nopl <sig>(%rip). */
+		".byte 0x0f, 0x1f, 0x05\n\t"
+		".long 0x53053053\n\t"	/* RSEQ_FLAGS */
+		"4:\n\t"
+		"jmp abort\n\t"
+		".popsection\n\t"
+		: /* gcc asm goto does not allow outputs */
+	       	: [cpu_id]              "r" (cpu),
+		[current_cpu_id]      "m" (__rseq_abi.cpu_id),
+		[rseq_cs]             "m" (__rseq_abi.rseq_cs),
+		/* final store input */
+		[v]                   "m" (*v),
+		[count]               "er" (count)
+		: "memory", "cc", "rax"
+		: abort
+	);
+	/* clang-format on */
+
+	return 0;
+abort:
+	return -1;
+}
+#endif
+int main(int argc, char *argv[])
+{
+#ifdef TESTRSEQ
+	int cpu, ret;
+#endif
+	intptr_t *cpu_data;
+	long nr_cpus = sysconf(_SC_NPROCESSORS_ONLN);
+
+	test_init(argc, argv);
+
+	cpu_data = calloc(nr_cpus, sizeof(*cpu_data));
+	if (!cpu_data) {
+		fail("calloc");
+		exit(EXIT_FAILURE);
+	}
+	register_thread();
+
+	test_msg("Ready for C/R\n");
+
+	test_daemon();
+	test_waitsig();
+#ifdef TESTRSEQ
+	check_thread();
+
+	cpu = RSEQ_ACCESS_ONCE(__rseq_abi.cpu_id_start);
+	ret = rseq_addv(&cpu_data[cpu], 2, cpu);
+	if (ret)
+		fail("Failed to increment per-cpu counter");
+	else
+		test_msg("cpu_data[%d] == %ld\n", cpu, (long int)cpu_data[cpu]);
+
+	/* TODO: remove this :) */
+	(void)cpu;
+	(void)ret;
+	(void)rseq_addv;
+	//unregister_thread();
+	(void)unregister_thread;
+	(void)register_thread;
+#endif
+	test_msg("cpu, cpu_data %d %ld\n", cpu, (long int)cpu_data[cpu]);
+	if (cpu_data[cpu] == 2)
+		pass();
+	else
+		fail();
+	//fail();
+	//unregister_thread();
+	//(void)unregister_thread;
+
+	return 0;
+}
+
+#else
```

Comment:
It is not very clear what this `#else` is for. Would it be possible to add a comment? For example, something like:
```C
#else  /* #if defined(__x86_64__) */
```